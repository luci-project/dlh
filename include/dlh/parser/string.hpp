#pragma once

#include <dlh/container/vector.hpp>
#include <dlh/types.hpp>

namespace Parser {

bool string(unsigned long long & target, const char * value);

bool string(long long & target, const char * value);

bool string(unsigned long & target, const char * value);

bool string(long & target, const char * value);

bool string(unsigned int & target, const char * value);

bool string(int & target, const char * value);

bool string(unsigned short & target, const char * value);

bool string(short & target, const char * value);

bool string(unsigned char & target, const char * value);

bool string(char & target, const char * value);

bool string(bool & target, const char * value);

bool string(const char * & target, const char * value);

template<typename T>
bool string(Vector<T> & target, const char * value) {
	T tmp;
	bool r = string(tmp, value);
	target.push_back(tmp);
	return r;
}

}  // namespace Parse
