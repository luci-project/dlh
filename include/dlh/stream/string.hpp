// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/stream/buffer.hpp>

template<size_t BUFFERSZ>
class StringStream : public BufferStream {
 protected:
	char buffer[BUFFERSZ];

 public:
	/*! \brief Default constructor  */
	StringStream() : BufferStream(buffer, BUFFERSZ) {}

	/*! \brief Destructor */
	virtual ~StringStream() {}
};
