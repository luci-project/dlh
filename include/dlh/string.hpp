#pragma once

#include <dlh/types.hpp>
#include <dlh/container/vector.hpp>

namespace String {

/*! \brief Find the first occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to first occurrence of the character
 *         or to null byte at the end of the string if not found
 */
const char* find_or_end(const char *s, int c);

/*! \brief Find the first occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to first occurrence of the needle
 *         or to null byte at the end of the string if not found
 */
const char* find_or_end(const char *haystack, const char *needle);

/*! \brief Find the first occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to first occurrence of the character
 *         or to nullptr if not found
 */
const char* find(const char *s, int c);

/*! \brief Find the first occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to first occurrence of the needle
 *         or to nullptr if not found
 */
const char* find(const char *haystack, const char *needle);

/*! \brief Find the last occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to last occurrence of the character
 *         or to nullptr if not found
 */
const char* find_last(const char *s, int c);

/*! \brief Find the last occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to last occurrence of the needle
 *         or to nullptr if not found
 */
const char* find_last(const char *haystack, const char *needle);

/*! \brief Replace characters in a string
 * \param target pointer to string (target memory will be modified!)
 * \param from character to be replaced
 * \param to replacement character
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 */
char * replace_inplace(char *target, int from, int to, size_t max = SIZE_MAX);

/*! \brief Replace characters in a string
 * \param target pointer to string (string will be copied)
 * \param from character to be replaced
 * \param to replacement character
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 * \note Target pointer has to be freed after usage
 */
char * replace(const char *target, int from, int to, size_t max = SIZE_MAX);

/*! \brief Replace stubstrings in a string
 * \param target pointer to string  (target memory will be modified!)
 * \param from substring to be replaced
 * \param to replacement substring (must be equal or shorter than `from`)
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 */
char * replace_inplace(char *target, const char *from, const char * to, size_t max = SIZE_MAX);

/*! \brief Replace stubstrings in a string
 * \param target pointer to string (string will be copied)
 * \param from substring to be replaced
 * \param to replacement substring
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 * \note Target pointer has to be freed after usage
 */
char * replace(const char *target, const char *from, const char * to, size_t max = SIZE_MAX);

/*! \brief Compare two strings
 * \param s1 first string
 * \param s2 second string
 * \return an integer less than, equal to, or greater than zero if first string is found, respectively,
 *         to be less than, to match, or be greater than second string
 */
int compare(const char *s1, const char *s2);

/*! \brief Compare two strings
 * \param s1 first string
 * \param s2 second string
 * \param n number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first string are
 *          found, respectively, to be less than, to match, or be greater than second string
 */
int compare(const char *s1, const char *s2, size_t n);

/*! \brief Calculate the length of a string
 * \param s pointer to a string
 * \return number of bytes in the string
 */
size_t len(const char *s);

/*! \brief Copy the contents of a string
 * including the terminating null byte (`\0`)
 * \param dest destination string buffer
 * \param src source string buffer
 * \return a pointer to the destination string buffer
 * \note Beware of buffer overruns!
 */
char* copy(char * dest, const char * src);  //NOLINT

/*! \brief Copy the contents of a string up to a maximum length
 * or the terminating null byte (`\0`), whatever comes first.
 * \param dest destination string buffer
 * \param src source string buffer
 * \param n maximum number of bytes to copy
 * \return a pointer to the destination string buffer
 * \note If there is no null byte (`\0`) among the first `n` bytes, the destination will not be null-terminated!
 */
char* copy(char * dest, const char * src, size_t n);

/*! \brief Duplicate a string
 * \param s pointer to a string
 * \return pointer to a duplicated string allocated with malloc or `nullptr` if allocation failed.
 */
char* duplicate(const char *s);

/*! \brief Duplicate a string
 * \param s pointer to a string
 * \param n maximum length
 * \return pointer to a duplicated string (up to maximum length, always null-terminated),
 *         allocated with malloc or `nullptr` if allocation failed.
 */
char* duplicate(const char *s, size_t n);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string (target memory will be modified!)
 * \param delimiter character to split string (will be overwritten with NULL character)
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 */
Vector<const char *> split_inplace(char * source, int delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string
 * \param delimiter character to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 * \note You have to free each element of the result vector!
 */
Vector<const char *> split(const char * source, int delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string (target memory will be modified!)
 * \param delimiter substring to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 */
Vector<const char *> split_inplace(char * source, const char * delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string
 * \param delimiter substring to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 * \note You have to free each element of the result vector!
 */
Vector<const char *> split(const char * source, const char * delimiter, size_t max = SIZE_MAX);
}  // namespace String
