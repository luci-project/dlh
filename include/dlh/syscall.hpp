#pragma once

#include <dlh/error.hpp>
#include <dlh/types.hpp>
#include <dlh/systypes.hpp>

// Handy helpers for errors
void warn(const char * msg);
void warn(const char * msg, const char * detail);
[[noreturn]] void die(const char * msg = nullptr, const char * detail = nullptr);

/*! \brief Syscalls with errno in return structure */
namespace Syscall {

/*! \brief Return Value with Errno */
template<typename T>
class ReturnValue {
	T _v;
	Errno _e;

 public:
	ReturnValue() : _e(ENONE) {}
	ReturnValue(Errno e) :  _v((T)(-1)), _e(e) {}
	ReturnValue(T v, Errno e = ENONE) :  _v(v), _e(e) {}
	ReturnValue(const ReturnValue& other) : _v(other._v), _e(other._e) {}
	ReturnValue(ReturnValue&& other) = default;

	inline T* operator->() { return &_v; }
	inline const T* operator->() const { return &_v; }
	inline T& operator*() & { return _v; }
	inline const T& operator*() const & { return _v; }
	inline bool success() const { return _e == ENONE; }
	inline bool failed() const { return _e != ENONE; }
	inline Errno error() const { return _e; }
	inline const char * error_message() const { return __errno_string(_e); }
	inline void warn_on_error(const char * msg = nullptr) const { if (!success()) warn(msg, error_message()); }
	inline void exit_on_error(const char * msg = nullptr) const { if (!success()) die(msg, error_message()); }
	inline void die_on_error(const char * msg = nullptr) const { if (!success()) die(msg, error_message()); }
	inline T& value() & { return _v; }
	inline const T& value() const & { return _v; }
	inline T& value_or_exit(const char * msg = nullptr) & { exit_on_error(msg); return value(); }
	inline const T& value_or_exit(const char * msg = nullptr) const & { exit_on_error(msg); return value(); }
	inline T& value_or_die(const char * msg = nullptr) & { die_on_error(msg); return value(); }
	inline const T& value_or_die(const char * msg = nullptr) const & { die_on_error(msg); return value(); }
	inline T& value_or_default(const T& v) & { return success() ? value() : v; }
	inline const T& value_or_default(const T& v) const & { return success() ? value() : v; }
	inline operator bool() const { return success(); }
	template<typename O> ReturnValue & operator=(const ReturnValue<O>& other) { _e = other._e; _v = other._v; return *this; }
	ReturnValue & operator=(const T& other) { _v = other; return *this; }
	template<typename O> bool operator==(const ReturnValue<O>& other) const { return _e == other._e && _v == other._v; }
	template<typename O> bool operator==(const O& other) const { return _e == ENONE && _v == other; }
	template<typename O> bool operator!=(const ReturnValue<O>& other) const { return _e != other._e && _v != other._v; }
	template<typename O> bool operator!=(const O& other) const { return _e != ENONE || _v != other; }
};


ReturnValue<int> nanosleep(const struct timespec *req, struct timespec *rem);
ReturnValue<unsigned> sleep(unsigned seconds);

pid_t gettid();
pid_t getpid();
pid_t getppid();

ReturnValue<int> getrlimit(rlimit_t resource, struct rlimit *rlim);
ReturnValue<int> arch_prctl(arch_code_t code, unsigned long addr);
ReturnValue<int> prctl(prctl_t option, unsigned long arg2, unsigned long arg3 = 0, unsigned long arg4 = 0, unsigned long arg5 = 0);

ReturnValue<int> sigaction(int sig, const struct sigaction * __restrict__ sa, struct sigaction * __restrict__ old);
ReturnValue<int> raise(signal_t sig);

[[noreturn]] void abort();
[[noreturn]] void crash();
[[noreturn]] void exit(int code);

ReturnValue<uintptr_t> mmap(uintptr_t start, size_t len, int prot, int flags, int fd, long off);
ReturnValue<uintptr_t> mremap(uintptr_t old_addr, size_t old_len, size_t new_len, int flags, uintptr_t new_addr = 0);
ReturnValue<int> mprotect(uintptr_t start, size_t len, int prot);
ReturnValue<int> munmap(uintptr_t start, size_t len);
ReturnValue<int> msync(uintptr_t start, size_t len, int flags);

ReturnValue<int> access(const char *filename, int amode);
ReturnValue<ssize_t> readlink(const char * __restrict__ path, char * __restrict__ buf, size_t bufsize);
ReturnValue<int> open(const char *filename, int flags, int mode = 0);
ReturnValue<ssize_t> read(int fd, void *buf, size_t count);
ReturnValue<ssize_t> write(int fd, const void *buf, size_t size);
ReturnValue<int> close(int fd);
ReturnValue<int> fcntl(int fd, fcntl_cmd_t cmd, unsigned long arg = 0);
ReturnValue<int> fallocate(int fd, int mode, off_t base, off_t len);
ReturnValue<int> ftruncate(int fd, off_t length);

ReturnValue<int> fstat(int fd, struct stat *st);
ReturnValue<int> stat(const char * __restrict__ path, struct stat * __restrict__ buf);
ReturnValue<int> lstat(const char * __restrict__ path, struct stat * __restrict__ buf);

ReturnValue<int> userfaultfd(int flags);
ReturnValue<int> ioctl(int fd, int req, void * arg = nullptr);
ReturnValue<int> memfd_create(const char *name, unsigned flags);

ReturnValue<int> inotify_init(int flags = 0);
ReturnValue<int> inotify_add_watch(int fd, const char *pathname, uint32_t mask);
ReturnValue<int> inotify_rm_watch(int fd, int wd);

ReturnValue<int> futex(int * __restrict__ uaddr, futex_op_t futex_op, int val, const void * __restrict__ timeout, int * __restrict__ uaddr2, int val3);

ReturnValue<int> brk(uintptr_t addr);
ReturnValue<uintptr_t> sbrk(intptr_t inc);
}  // namespace Syscall
