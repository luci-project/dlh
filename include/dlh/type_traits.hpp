#pragma once

extern "C++" {

template<typename T> struct remove_cv                   { typedef T type; };
template<typename T> struct remove_cv<const T>          { typedef T type; };
template<typename T> struct remove_cv<volatile T>       { typedef T type; };
template<typename T> struct remove_cv<const volatile T> { typedef T type; };

template<typename T> struct remove_const                { typedef T type; };
template<typename T> struct remove_const<const T>       { typedef T type; };

template<typename T> struct remove_volatile             { typedef T type; };
template<typename T> struct remove_volatile<volatile T> { typedef T type; };

template<typename T> struct remove_reference            { typedef T type; };
template<typename T> struct remove_reference<T&>        { typedef T type; };
template<typename T> struct remove_reference<T&&>       { typedef T type; };

template<typename T, T v>
struct integral_constant {
    static constexpr T value = v;
    using value_type = T;
    using type = integral_constant;
    constexpr operator value_type() const noexcept { return value; }
    constexpr value_type operator()() const noexcept { return value; }
};

typedef integral_constant<bool, true>  true_type;
typedef integral_constant<bool, false> false_type;

template<typename T> struct is_const          : false_type {};
template<typename T> struct is_const<const T> : true_type {};

template<typename T> struct is_reference      : false_type {};
template<typename T> struct is_reference<T&>  : true_type {};
template<typename T> struct is_reference<T&&> : true_type {};

template<typename>   struct is_lvalue_reference      : false_type {};
template<typename T> struct is_lvalue_reference<T&>  : true_type {};

template<typename>   struct is_rvalue_reference      : false_type {};
template<typename T> struct is_rvalue_reference<T&&> : true_type {};

template<typename T> struct is_volatile             : false_type {};
template<typename T> struct is_volatile<volatile T> : true_type {};

template<typename> struct is_array                    : false_type { };
template<typename T, size_t S> struct is_array<T[S]>  : true_type { };
template<typename T> struct is_array<T[]>             : true_type { };

template<typename T, typename U> struct is_same       : false_type {};
template<typename T>             struct is_same<T, T> : true_type {};


template<typename> struct is_integral_base : false_type {};
template<> struct is_integral_base<bool> : true_type {};
template<> struct is_integral_base<char> : true_type {};
template<> struct is_integral_base<unsigned char> : true_type {};
template<> struct is_integral_base<short> : true_type {};
template<> struct is_integral_base<unsigned short> : true_type {};
template<> struct is_integral_base<int> : true_type {};
template<> struct is_integral_base<unsigned int> : true_type {};
template<> struct is_integral_base<long> : true_type {};
template<> struct is_integral_base<unsigned long> : true_type {};
template<> struct is_integral_base<long long> : true_type {};
template<> struct is_integral_base<unsigned long long> : true_type {};
template<typename T> struct is_integral : is_integral_base<remove_cv<T>> {};

template<typename T> inline const char * integral_name() { return "[complex type]"; };
template<> inline const char * integral_name<bool>() { return "bool"; };
template<> inline const char * integral_name<char>() { return "char"; };
template<> inline const char * integral_name<unsigned char>() { return "unsigned char"; };
template<> inline const char * integral_name<short>() { return "short"; };
template<> inline const char * integral_name<unsigned short>() { return "unsigned short"; };
template<> inline const char * integral_name<int>() { return "int"; };
template<> inline const char * integral_name<unsigned int>() { return "unsigned int"; };
template<> inline const char * integral_name<long>() { return "long"; };
template<> inline const char * integral_name<unsigned long>() { return "unsigned long"; };
template<> inline const char * integral_name<long long>() { return "long long"; };
template<> inline const char * integral_name<unsigned long long>() { return "unsigned long long"; };

}
