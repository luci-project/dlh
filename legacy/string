#pragma once

#include <dlh/strptr.hpp>

namespace std {

using string = StrPtr;

} // namespace std
