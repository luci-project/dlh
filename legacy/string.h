#pragma once

#include <dlh/string.hpp>
