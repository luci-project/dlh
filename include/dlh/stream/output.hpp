#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
