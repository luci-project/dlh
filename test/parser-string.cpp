// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#include <dlh/stream/output.hpp>
#include <dlh/parser/string.hpp>
#include <dlh/type_traits.hpp>
#include <dlh/syscall.hpp>

template<typename T>
void test(const char * str, T value) {
	T tmp;
	cout << "Parser::string(" << integral_name<T>() << ", \"" << str << "\") " << flush;
	if (Parser::string(tmp, str)) {
		cout << "= " << tmp << endl;
		if (tmp != value) {
			cerr << "Wrong value: " << tmp << " != " << value << endl;
			Syscall::exit(1);
		}
	} else {
		cerr << "failed!" << endl;
		Syscall::exit(1);
	}
}

template<typename T>
void testfail(const char * str) {
	T tmp;
	cout << "Parser::string(" << integral_name<T>() << ", \"" << str << "\") " << flush;
	if (Parser::string(tmp, str)) {
		cout << "= " << tmp << flush;
		cerr << " should fail!" << endl;
		Syscall::exit(1);
	} else {
		cout << "failed (as expected)!" << endl;
	}
}

int main(int argc, const char *argv[]) {
	(void) argc;
	(void) argv;

	test<int>("0", 0);
	test<int>("+0", 0);
	test<int>("-0", 0);
	test<unsigned>("0", 0);
	test<unsigned>("+0", 0);

	test<int>("42", 42);
	test<int>("+42", 42);
	test<int>("+ 42", 42);
	test<int>(" + 42", 42);
	test<int>(" +42", 42);
	test<int>("-42", -42);
	test<int>(" - 42", -42);
	test<int>("123'000'000", 123'000'000);
	test<int>("123,000,000", 123'000'000);
	test<int>("123 000 000", 123'000'000);
	testfail<int>("123-000");
	testfail<int>("123'000,000");
	testfail<int>("123.000.000");


	test<unsigned>("42", 42);
	test<unsigned>("+42", 42);
	test<unsigned>("+ 42", 42);
	test<unsigned>(" + 42", 42);
	testfail<unsigned>("-42");
	testfail<unsigned>(" - 42");
	test<unsigned>("123'000'000", 123'000'000);
	test<unsigned>("123,000,000", 123'000'000);
	test<unsigned>("123 000 000", 123'000'000);

	test<int>("023", 023);
	test<int>("+023", 023);
	test<int>(" + 023", 023);
	test<int>("-023", -023);
	test<int>(" - 023", -023);

	test<int>("0x23", 0x23);
	test<int>("+0x23", 0x23);
	test<int>(" + 0x23", 0x23);
	test<int>("-0x23", -0x23);
	test<int>(" - 0x23", -0x23);

	test<int>("0XAFfe", 0xaffe);
	testfail<int>("affe");
	testfail<int>("xaffe");

	test<int>("0b10100101", 0b10100101);
	test<int>("0B1101", 0b1101);
	testfail<int>("b1101");

	test<int>("023", 023);
	test<int>("0x23", 0x23);

	test<int8_t>("0x7f", 0x7f);
	test<int8_t>("-0x80", -0x80);
	testfail<int8_t>("0x80");
	testfail<int8_t>("-0x81");
	test<uint8_t>("0xff", 0xff);
	testfail<uint8_t>("0x100");

	test<int16_t>("0x7fff", 0x7fff);
	test<int16_t>("-0x8000", -0x8000);
	testfail<int16_t>("0x8000");
	testfail<int16_t>("-0x8001");
	test<uint16_t>("0xffff", 0xffff);
	testfail<uint16_t>("0x10000");

	test<int32_t>("0x7fffffff", 0x7fffffff);
	test<int32_t>("-0x80000000", -0x80000000);  // NOLINT
	testfail<int32_t>("0x80000000");
	testfail<int32_t>("-0x80000001");
	test<uint32_t>("0xffffffff", 0xffffffff);
	testfail<uint32_t>("0x100000000");

	test<int64_t>("0x7fffffffffffffff", 0x7fffffffffffffffL);
	test<int64_t>("-0x8000000000000000", -0x8000000000000000L);  // NOLINT
	testfail<int64_t>("0x8000000000000000");
	testfail<int64_t>("-0x8000000000000001");
	test<uint64_t>("0xffffffffffffffff", 0xffffffffffffffffUL);
	testfail<uint64_t>("0x10000000000000000");

	test<bool>("f", false);
	test<bool>("False", false);
	test<bool>("0", false);
	test<bool>("no", false);
	test<bool>("Nope", false);

	test<bool>("t", true);
	test<bool>("True", true);
	test<bool>("1", true);
	test<bool>("yes", true);
	test<bool>("Yay", true);

	return 0;
}
