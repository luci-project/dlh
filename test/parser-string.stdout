Parser::string(int, "0") = 0
Parser::string(int, "+0") = 0
Parser::string(int, "-0") = 0
Parser::string(unsigned int, "0") = 0
Parser::string(unsigned int, "+0") = 0
Parser::string(int, "42") = 42
Parser::string(int, "+42") = 42
Parser::string(int, "+ 42") = 42
Parser::string(int, " + 42") = 42
Parser::string(int, " +42") = 42
Parser::string(int, "-42") = -42
Parser::string(int, " - 42") = -42
Parser::string(int, "123'000'000") = 123000000
Parser::string(int, "123,000,000") = 123000000
Parser::string(int, "123 000 000") = 123000000
Parser::string(int, "123-000") failed (as expected)!
Parser::string(int, "123'000,000") failed (as expected)!
Parser::string(int, "123.000.000") failed (as expected)!
Parser::string(unsigned int, "42") = 42
Parser::string(unsigned int, "+42") = 42
Parser::string(unsigned int, "+ 42") = 42
Parser::string(unsigned int, " + 42") = 42
Parser::string(unsigned int, "-42") failed (as expected)!
Parser::string(unsigned int, " - 42") failed (as expected)!
Parser::string(unsigned int, "123'000'000") = 123000000
Parser::string(unsigned int, "123,000,000") = 123000000
Parser::string(unsigned int, "123 000 000") = 123000000
Parser::string(int, "023") = 19
Parser::string(int, "+023") = 19
Parser::string(int, " + 023") = 19
Parser::string(int, "-023") = -19
Parser::string(int, " - 023") = -19
Parser::string(int, "0x23") = 35
Parser::string(int, "+0x23") = 35
Parser::string(int, " + 0x23") = 35
Parser::string(int, "-0x23") = -35
Parser::string(int, " - 0x23") = -35
Parser::string(int, "0XAFfe") = 45054
Parser::string(int, "affe") failed (as expected)!
Parser::string(int, "xaffe") failed (as expected)!
Parser::string(int, "0b10100101") = 165
Parser::string(int, "0B1101") = 13
Parser::string(int, "b1101") failed (as expected)!
Parser::string(int, "023") = 19
Parser::string(int, "0x23") = 35
Parser::string(char, "0x7f") = 
Parser::string(char, "-0x80") = €
Parser::string(char, "0x80") failed (as expected)!
Parser::string(char, "-0x81") failed (as expected)!
Parser::string(unsigned char, "0xff") = ÿ
Parser::string(unsigned char, "0x100") failed (as expected)!
Parser::string(short, "0x7fff") = 32767
Parser::string(short, "-0x8000") = -32768
Parser::string(short, "0x8000") failed (as expected)!
Parser::string(short, "-0x8001") failed (as expected)!
Parser::string(unsigned short, "0xffff") = 65535
Parser::string(unsigned short, "0x10000") failed (as expected)!
Parser::string(int, "0x7fffffff") = 2147483647
Parser::string(int, "-0x80000000") = -2147483648
Parser::string(int, "0x80000000") failed (as expected)!
Parser::string(int, "-0x80000001") failed (as expected)!
Parser::string(unsigned int, "0xffffffff") = 4294967295
Parser::string(unsigned int, "0x100000000") failed (as expected)!
Parser::string(long, "0x7fffffffffffffff") = 9223372036854775807
Parser::string(long, "-0x8000000000000000") = -9223372036854775808
Parser::string(long, "0x8000000000000000") failed (as expected)!
Parser::string(long, "-0x8000000000000001") failed (as expected)!
Parser::string(unsigned long, "0xffffffffffffffff") = 18446744073709551615
Parser::string(unsigned long, "0x10000000000000000") failed (as expected)!
Parser::string(bool, "f") = false
Parser::string(bool, "False") = false
Parser::string(bool, "0") = false
Parser::string(bool, "no") = false
Parser::string(bool, "Nope") = false
Parser::string(bool, "t") = true
Parser::string(bool, "True") = true
Parser::string(bool, "1") = true
Parser::string(bool, "yes") = true
Parser::string(bool, "Yay") = true
